src/ring_theory/polynomial/basic.lean

```diff
@@ -215,6 +215,72 @@ begin
   refl,
 end
 
+lemma geom_sum_X_comp_X_add_one_eq_sum (n : ℕ) :
```

Comment:
Here's a slightly different lemma:
```
lemma geom_sum_X_comp_X_add_one_eq_sum (n : ℕ) :
  (geom_sum (X : R[X]) n).comp (X + 1) =
  (finset.range n).sum (λ (i : ℕ), (n.choose (i + 1) : R[X]) * X ^ i) :=
begin
  ext i,
  transitivity (n.choose (i + 1) : R), swap,
  { simp only [finset_sum_coeff, ← C_eq_nat_cast, coeff_C_mul_X_pow],
    rw [finset.sum_eq_single i, if_pos rfl],
    { simp only [@eq_comm _ i, if_false, eq_self_iff_true, implies_true_iff] {contextual := tt}, },
    { simp only [nat.lt_add_one_iff, nat.choose_eq_zero_of_lt, nat.cast_zero, finset.mem_range,
        not_lt, eq_self_iff_true, if_true, implies_true_iff] {contextual := tt}, } },
  induction n with n ih generalizing i,
  { simp only [geom_sum_zero, zero_comp, coeff_zero, nat.choose_zero_succ, nat.cast_zero], },
  simp only [geom_sum_succ', ih, add_comp, pow_comp, X_comp, coeff_add, nat.choose_succ_succ,
    nat.cast_add, add_pow, one_pow, mul_one, finset_sum_coeff, ← C_eq_nat_cast, mul_comm _ (C _),
    coeff_C_mul_X_pow],
  rw [finset.sum_eq_single i, if_pos rfl],
  { simp only [@eq_comm _ i, if_false, eq_self_iff_true, implies_true_iff] {contextual := tt}, },
  { simp only [nat.lt_add_one_iff, nat.choose_eq_zero_of_lt, nat.cast_zero, finset.mem_range,
      eq_self_iff_true, if_true, implies_true_iff, not_le] {contextual := tt}, },
end
```
The proof can be further deduplicated, if we have a better way of handling the RHS (and in particular the coefficients of the RHS).